File: packages/ruby-sdk/lib/opensend/http_client.rb
# frozen_string_literal: true

require "json"
require "net/http"
require "uri"

module OpenSend
  # Low-level HTTP client shared by every resource. Supports GET, POST, PATCH,
  # and DELETE. Raises OpenSend::Error on non-2xx responses.
  class HttpClient
    def initialize(api_key:, base_uri:)
      @api_key = api_key
      @base_uri = base_uri
    end

    # GET with optional Hash of query params.
    def get(path, params = {})
      uri = endpoint(path, params)
      request = build_request(Net::HTTP::Get, uri)
      perform(uri, request)
    end

    # POST with a body Hash (serialised to JSON).
    def post(path, payload = nil, headers: {})
      uri = endpoint(path)
      request = build_request(Net::HTTP::Post, uri, headers: headers)
      request.body = payload.nil? ? "" : JSON.generate(payload)
      perform(uri, request)
    end

    # PATCH with a body Hash.
    def patch(path, payload = nil)
      uri = endpoint(path)
      request = build_request(Net::HTTP::Patch, uri)
      request.body = payload.nil? ? "" : JSON.generate(payload)
      perform(uri, request)
    end

    # DELETE (no body).
    def delete(path)
      uri = endpoint(path)
      request = build_request(Net::HTTP::Delete, uri)
      perform(uri, request)
    end

    private

    def endpoint(path, params = {})
      uri = @base_uri.dup
      base_path = uri.path.to_s.sub(%r{/+\z}, "")
      request_path = "/#{path.to_s.sub(%r{\A/+}, "")}"
      uri.path = base_path.empty? ? request_path : "#{base_path}#{request_path}"
      uri.fragment = nil
      if params && !params.empty?
        uri.query = URI.encode_www_form(params.reject { |_, v| v.nil? })
      else
        uri.query = nil
      end
      uri
    end

    def build_request(klass, uri, headers: {})
      request = klass.new(uri)
      request["Authorization"] = "Bearer #{@api_key}"
      request["Content-Type"] = "application/json"
      request["Accept"] = "application/json"
      request["User-Agent"] = USER_AGENT
      headers.each { |k, v| request[k] = v }
      request
    end

    def perform(uri, request)
      response = Net::HTTP.start(uri.host, uri.port, use_ssl: uri.scheme == "https") do |http|
        http.request(request)
      end

      body = response.body.to_s

      unless response.is_a?(Net::HTTPSuccess)
        raise parse_error(response, body)
      end

      body.empty? ? {} : JSON.parse(body)
    rescue JSON::ParserError => e
      raise Error.new(
        "Invalid JSON response from OpenSend",
        status_code: 0,
        name: "parse_error",
        code: "parse_error",
        body: e.message
      )
    rescue IOError, SystemCallError, Timeout::Error, SocketError => e
      raise Error.new(
        e.message,
        status_code: 0,
        name: "request_error",
        code: "request_error"
      )
    end

    def parse_error(response, body)
      envelope = begin
        parsed = body.empty? ? {} : JSON.parse(body)
        parsed.is_a?(Hash) ? parsed : {}
      rescue JSON::ParserError
        {}
      end

      message = envelope["message"] || envelope["error"] || response.message || "OpenSend API request failed"

      Error.new(
        message,
        status_code: response.code.to_i,
        name: envelope["name"],
        code: envelope["code"],
        details: envelope["details"],
        body: body
      )
    end
  end
end

File: packages/php-sdk/src/Http/HttpClient.php
<?php

declare(strict_types=1);

namespace OpenSend\Http;

use JsonException;
use OpenSend\Constants;
use OpenSend\Contracts\HttpTransport;
use OpenSend\Errors\ApiException;
use OpenSend\ValueObjects\RequestOptions;

final class HttpClient
{
    public function __construct(
        private readonly string $apiKey,
        private readonly string $baseUrl,
        private readonly HttpTransport $transport = new StreamHttpTransport(),
    ) {
    }

    /**
     * @param array<string, mixed>|null $payload
     * @return array<string, mixed>
     */
    public function request(string $method, string $path, ?array $payload = null, ?RequestOptions $options = null): array
    {
        $body = null;
        if ($payload !== null) {
            try {
                $body = json_encode($payload, JSON_THROW_ON_ERROR);
            } catch (JsonException $exception) {
                throw new ApiException('Failed to encode OpenSend request JSON.', 0, 'request_error', 'request_error', previous: $exception);
            }
        }

        $headers = [
            'Accept' => 'application/json',
            'Authorization' => 'Bearer ' . $this->apiKey,
            'Content-Type' => 'application/json',
            'User-Agent' => Constants::USER_AGENT,
        ];

        if ($options?->idempotencyKey !== null) {
            $headers['Idempotency-Key'] = $options->idempotencyKey;
        }

        $response = $this->transport->send(
            strtoupper($method),
            $this->endpoint($path),
            $headers,
            $body,
        );

        $decoded = $this->decodeJson($response->body, $response->statusCode);
        if ($response->statusCode < 200 || $response->statusCode >= 300) {
            throw $this->apiException($response, $decoded);
        }

        return $decoded;
    }

    private function endpoint(string $path): string
    {
        return rtrim($this->baseUrl, '/') . '/' . ltrim($path, '/');
    }

    /**
     * @return array<string, mixed>
     */
    private function decodeJson(string $body, int $statusCode): array
    {
        if (trim($body) === '') {
            return [];
        }

        try {
            $decoded = json_decode($body, true, 512, JSON_THROW_ON_ERROR);
        } catch (JsonException $exception) {
            throw new ApiException(
                'Invalid JSON response from OpenSend.',
                $statusCode,
                'parse_error',
                'parse_error',
                rawBody: $body,
                previous: $exception,
            );
        }

        if (!is_array($decoded)) {
            throw new ApiException(
                'OpenSend returned a JSON response that was not an object.',
                $statusCode,
                'parse_error',
                'parse_error',
                rawBody: $body,
            );
        }

        /** @var array<string, mixed> $decoded */
        return $decoded;
    }

    /**
     * @param array<string, mixed> $decoded
     */
    private function apiException(HttpResponse $response, array $decoded): ApiException
    {
        $message = $this->stringValue($decoded['message'] ?? null)
            ?? $this->stringValue($decoded['error'] ?? null)
            ?? 'OpenSend API request failed.';
        $name = $this->stringValue($decoded['name'] ?? null);
        $code = $this->stringValue($decoded['code'] ?? null);
        $details = isset($decoded['details']) && is_array($decoded['details']) ? $decoded['details'] : null;

        return new ApiException($message, $response->statusCode, $name, $code, $details, $response->body);
    }

    private function stringValue(mixed $value): ?string
    {
        return is_string($value) && $value !== '' ? $value : null;
    }
}
